Add detect_gnu_libc() and detect_musl_libc() functions

jwillikers opened the pull request:
Changelog: Feature: Add `detect_libc` to the `detect_api` to get the name and version of the C library.
Docs: https://github.com/conan-io/docs/pull/3590

This PR adds a simple function to detect the version of glibc to the detect_api. This functionality is only intended for Linux systems using glibc. The function is detect_gnu_libc().
It simply runs ldd --version to detect the version of glibc. It is also possible to detect the version of the libc library by running the libc.so file. The location of this library varies much more than the path to ldd, so ldd --version is used instead. The path to ldd is passed as an argument, /usr/bin/ldd by default. This can modified to suite alternate sysroots.
It can even be used in cross-compilation scenarios where emulation is available to execute ldd.

The detect_gnu_libc function does a simple sanity check to verify that glibc is indeed being used. This avoids confusion on systems using an alternative libc implementation, i.e. musl.

I also added a function to detect the version of the musl libc library as well. This function is detect_musl_libc() and works the same way.

Addresses conan-io/docs#3960, at least in part.



- [x] Refer to the issue that supports this Pull Request.
- [x] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [x] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one.

jwillikers committed:
Add detect_gnu_libc() and detect_musl_libc() functions

This PR adds a simple function to detect the version of glibc to the detect_api.
This functionality is only intended for Linux systems using glibc.
The function is detect_gnu_libc().
It simply runs ldd --version to detect the version of glibc.
It is also possible to detect the version of the libc library by running the libc.so file.
The location of this library varies much more than the path to ldd, so ldd --version is used instead.
The path to ldd is passed as an argument, /usr/bin/ldd by default.
This can modified to suite alternate sysroots.
It can even be used in cross-compilation scenarios where emulation is available to execute ldd.

The detect_gnu_libc function does a simple sanity check to verify that glibc is indeed being used.
This avoids confusion on systems using an alternative libc implementation, i.e. musl.

I also added a function to detect the version of the musl libc library as well.
This function is detect_musl_libc() and works the same way.

Addresses #3972, at least in part.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/conan-io/conan?pullRequest=15683) <br/>All committers have signed the CLA.

jwillikers commented:
Not sure the best way to add some automated unit tests for these functions, particularly since they are only applicable under Linux and need to run on a Linux system, unless I mock out `ldd`, which would probably be fine for unit tests.

memsharded commented on conan/internal/api/detect_api.py:
```diff
@@ -122,6 +122,47 @@ def _get_e2k_architecture():
+        return None
+
+
+def detect_musl_libc(ldd="/usr/bin/ldd"):
```
Functionally looks good, but it would be necessary to understand the usage patterns, which can be made explicit in a test.

Maybe it is better just a single function that returns ``libc, version = detect_api.detect_libc()``, so ``libc`` gets the value ``glibc`` or ``musl``?

jwillikers replied:
That's definitely possible and seems useful. I think it might imply the scope applies to more than just Linux, however, since macOS, BSD*, and Windows are all going to have their own libc implementation. I guess the functionality could always be expanded in the future, but it is quite nice to be able to pass `ldd` as an argument, too. I'll work on getting the tests sorted out tomorrow.

memsharded replied:
Yes, exactly, I think it can be expanded in the future in different forms like ``detect_api.detect_osx_libc()``, via a new argument, or just by adding functionality if we are sure it will not break.

jwillikers replied:
Alright, I've added a `detect_libc` function to be used instead of ones for the individual C libraries which returns a tuple of the name, i.e. `gnu` or `musl`, and the version.

I broke out the actual parsing logic into separate functions and added unit tests for this logic using the output of the `ldd` command.

I added a functional test for the `detect_libc` function.

memsharded reviewed (commented):
Thanks very much for your contribution @jwillikers 

It is true that testing these is typically not easy.

I would test it very similarly to ``TestProfileDetectAPI`` in "functional" tests, without a strict validation of the specific return version, just checking that it doesn't crash and return a version is fine. For the ``musl`` that will not be in CI, checking that it returns None without crashing would be enough, as long as you test it locally and works for you.

jwillikers committed:
Break out parse functions and add detect_libc function

jwillikers committed:
Add tests and fix detection of musl libc

jwillikers committed:
Add functional tests for detect_gnu_libc and detect_musl_libc

These just check that the functions run without errors.
This avoids assumptions about what C library is being used.

jwillikers committed:
Unify on a single function

memsharded commented on conans/test/unittests/util/detect_libc_test.py:
```diff
@@ -0,0 +1,84 @@
+from conan.internal.api.detect_api import _parse_gnu_libc, _parse_musl_libc
+
+
+class DetectedLibcTest(unittest.TestCase):
```
We are trying not to use ``unittest`` anymore, but only ``pytest``. It is ok, no need to change it, just FYI for future.

jwillikers replied:
Oh, I don't mind. I've updated that.

memsharded commented on conan/internal/api/detect_api.py:
```diff
@@ -122,6 +122,93 @@ def _get_e2k_architecture():
+    ConanOutput(scope="detect_api").warning(
+        f"Couldn't detect the libc provider and version"
+    )
+    return None
```
Better return ``None, None``, it is easier for jinja templates to check, and not raise error if  used as ``lib, version = ...``

jwillikers replied:
Good catch, fixed!

memsharded reviewed (commented):
Looks good, just a minor detail about returning ``None`` instead of ``None, None``

jwillikers committed:
Replace unittest with pytest

jwillikers committed:
Fix return for detect_libc function

memsharded reviewed (approved):
Looks good, many thanks!
Merging this for next release 2.2.

I'd recommend documenting this as "experimental", we might discover cases we didn't think about it when other users use it.
